README.md
# nanoboard
nanoboard client + sources in a single zip: https://github.com/nanoboard/nanoboard/releases

Внимание! Статья устарела и нуждается в обновлении

За время существования статьи спецификации поменялись и дополнились такими элементами как капча и Proof Of Work...
## Причина создания

 - Модераторский произвол
 - Мечта о неубиваемой борде

## Ценности Наноборды

 - простота формата нанопостов: хэш + текст сообщения = хэш (на который могут ссылаться другие нанопосты)
 - отсутствие владельца, прав, уровней доступа, невозможность централизованной модерации или отключения наноборды
 - отсутствие стандартизированного протокола распространения, в самой простой форме нанопост можно скопировать и вставить как текст
 - паразитирование на существующих ресурсах в интернете и скрытом интернете вместо организации p2p-соединений и/или серверов (низкий уровень детектирования + эксплуатация чужих мощностей)
 - поощряется помощь участникам и наноборде в виде ретрансляции чужих нанопостов (кроме огромных)
 - клиенты для просмотра собирают все нанопосты воедино и формируют цепочки - так сохраняется логическая связь между нанопостами и отпадает необходимость во временных метках (подверженных фальсификации)
 - клиенты для просмотра должны давать участнику возможность легко фильтровать нежелательный мусор, чтобы не видеть его и не распространять
 - клиенты для просмотра облегчают участнику распространение своих и чужих нанопостов (кроме мусора)
 - клиенты для просмотра ограничивают распространение слишком огромных чужих нанопостов, чтобы пресекать злоупотребление "трафиком" наноборды

## Спецификации

Каждый нанопост обязательно ссылается на другой нанопост поредством хэш-кода.

Хэш код 16-байтовый и выглядит как 32-символьный кусочек текста в шестандцатиричном текстовом виде.

Рассмотрим пример на JavaScript.

Нам понадобится SHA-256, взять можно например отсюда: http://www.movable-type.co.uk/scripts/sha256.html

или отсюда: https://username1565.github.io/sha256/sha256_utf8.html

```
var post = "00000000000000000000000000000000{Welcome to Nanoboard}"
var hash = Sha256.hash(post).substring(0,32) // берём хэш f682830a470200d738d32c69e6c2b8a43b44435f48ba000cfb5203b22e9bf026 в hex и отбрасываем всё кроме первых 32 символов
console.log(hash) // результат: f682830a470200d738d32c69e6c2b8a4
```
Хэш считается с байтов utf-8 кодировки. В nboard этот пост и его хэш используются как "корневые" но это не обязательно.

Теперь попробуем "ответить" на корневой пост.

```
var answer = hash + "Проверка ответа."
console.log(answer) // результат: f682830a470200d738d32c69e6c2b8a4Проверка ответа.
```

Ответим снова.

```
function getHash(content) { return Sha256.hash(content).substring(0,32); }
var answer2 = getHash(answer) + "Второй уровень."
console.log(answer2) // результат: 77bb9e5d1a64facc551210862e2c6536Второй уровень.
```
Вот так выглядят нанопосты, естественно, внутри можно ставить переносы строк и любые utf-8 символы.

Теперь осталось только распространить их. По хэшам все связывается в цепочки. Сложно сказать что является тредом среди нанопостов.

Определенно если на нанопост существует несколько ответов то с него можно начинать просмотр "треда".

Начать разворачивание можно с любого уровня, это уже на плечах реализаций просмотрщиков Наноборды.

В nboard можно разворачивать любые нанопосты кроме корневого, а можно просто просматривать только сам нанопост

и его ответы (следующий уровень).
_______
Наноборда берет начало с Анонимных Имидж Борд, Имидж означает наличие картинок (image).

Хотя существуют чисто текстовые борды, наноборда такой не является, потому как внутрь можно вставлять base64-изображения.

В nboard это реализовано так: если в нанопосте встречается конструкция вида [img=.....]

(где вместо точек - base64 строчка), то всё это дело превращается в изображение.

Кроме того nboard поддерживает разметку:

**[b]жирный[/b]**, *[i]курсив[/i]*, ~~[s]перечеркнутый[/s]~~, [sp]спойлер[/sp], [u]подчёркнутый[/u], [g]зелёный[/g].

Также, возможно распространять в виде base64 - любые файлы (размером до 65535 байт).

В основном, это zip-архивы, [но могут быть и любые другие файлы](https://github.com/nanoboard/nanoboard/issues/6).

Файлы, могут быть прекриплены, как JPEG-картинка, без сжатия, тогда, в нанопосте, используется:

 - либо старая конструкция [xmg=.....], где вместо точек - base64 несжатого файла,

   В этом случае, когда скрипт наноборды находит тег xmg, он автоматически добавляет ссылку на закачку zip-архива,

   превращая её в dataURL.

 - либо же используется bb-код file, с именем и типом файла - как у карасика.
 
   Тогда, при парсинге, имя файла попадает в ссылку, в атрибут download, а тип файла - в dataURL,
   
   и бинарный файл - можно закачать, по ссылке.

## PNG-контейнеры

На АИБ обычно можно свободно постить PNG-картинки. А в PNG-картинку можно вшить данные.

В nboard используется подход (и, собственно, код) из этой статьи: http://blog.andersen.im/2014/11/hiding-your-bits-in-the-bytes/ 

Вкратце суть в том что информация размещается в нижних битах компонентов R, G и B каждого пикселя

и это совершенно незаметно на глаз.

Разница между 24битным цветом и 21битным цветом с дизерингом (информациионный шум), по сути - минимальна.

Размер картинки после этого обычно увеличивается но не разительно.

Информация хранится именно в пикселях, что позволяет скопировать картинку в буфер обмена,

вставить в графический редактор, сохранить, скажем, в BMP, а потом пересохранить в PNG - вшитая информация не пропадёт.

Итак nboard берет некоторое количество свежих (по дате поступления на компьютер пользователя) постов,

а также некоторое количество случайно выбранных постов за всё время.

При этом избегаются скрытые пользователем посты, а крупные посты отправляются только если являются свежими.

В [наноборде карасика](https://github.com/Karasiq/nanoboard) также имеется кнопка "В очередь" и "Из очереди",

позволяющая тонко настроить список очереди нанопостов, для упаковки их оттуда - в контейнер.

Случайные посты, можно также отключить там.
_______
Но это там...

Здесь же, упаковка массива постов происходит следующим образом:

```
    public static byte[] Pack(NanoPost[] posts)
    {
        List<byte> bytes = new List<byte>();
        // количество постов (от 000000 до ffffff) (строчка в utf-8 байтах):
        bytes.AddRange(Encoding.UTF8.GetBytes(posts.Length.ToString("x6"))); 

        foreach (var p in posts)
        {
            // количество символов (не байтов!) в нанопосте (включая родительский хэш):
            var len = p.SerializedString().Length; 
            // тот же формат: от 000000 до ffffff (строчка в utf-8 байтах):
            bytes.AddRange(Encoding.UTF8.GetBytes(len.ToString("x6"))); 
        }

        foreach (var p in posts)
        {
            // utf-8 байты каждого поста по очереди без разделителей:
            bytes.AddRange(p.SerializedBytes()); 
        }

        // сжимаем результат gzip-ом
        return GZipUtil.Compress(bytes.ToArray());
    }
```

Перед упаковкой в контейнер (после сжатия gzip-ом) применяется шифрование salsa20 по некоторому ключу.

Для Наноборды пока есть один общий ключ, но дальше можно отделяться в отдельные зашифрованные Наноборды

(что нежелательно но иногда может быть необходимо):

```
        // Параметры инициализации salsa20 для заданного ключа (key):
        byte[] initKey = SHA256.Create().ComputeHash(Encoding.UTF8.GetBytes(key));
        byte[] initVec = SHA256.Create().ComputeHash(Encoding.UTF8.GetBytes(key).Reverse().ToArray());
        initKey = initKey.Crop(32);
        initVec = initVec.Crop(8);
```
_______
На этом пока что всё.

Напомню, что PNG-контейнеры - [не единственный возможный способ](https://username1565.github.io/js-jpeg-steg/) обмена нанопостами

и различные АИБ - [не единственное возможное место](https://github.com/username1565/nanoboard/issues/2) для паразитирования.

Одна из экстремальных идей - запихивание нанопостов в QR-коды и разбрасывание/расклеивание этих QR-кодов в местах,

где их будет удобно считывать (и, собственно, наклеивать/разбрасывать).

Так Наноборда сможет продолжить существование даже без Интернета, например, в метро.

Разумеется, вариант почти фантастический, но пример демонстрирует гибкость и ту самую неубиваемость Наноборды.

Ведь используя один лишь смартфон, с приложением в виде сканера qr-кодов,

даже в метро можно декодировать из qr-кода - коротенькую magnet-ссылку,

вытащить по ней - пак пикч из облачного торрента,

распарсить эти picture в бункере, на нейрокомпьютере, и залезть в нанабороду.

Алсо, может быть привинчена - IRC-координация (а это технология, она неубиваема, ага),

[Tox](https://tox.chat/download.html) (а там - [ассиметричная криптография](https://username1565.github.io/js-nacl/generate_keypair_and_crypt.html) на библиотеке [NaCl](https://username1565.github.io/js-nacl/)),

p2p (а там вообще - тотальная децентрализация пиринговых сетей же), 

и многое другое... 

## nanodb.exe-source from nanoboard-restore.zip

nanodb.exe-source from https://github.com/nanoboard/nanoboard/releases/download/0/nanoboard-restore.zip
nanoboard-restore\nanodb.exe-source
Version: Client 3.0
Language: C#

Q: How to compile nanodb.exe?
A:
1. Win+R
2. cmd, enter.
3. disk:\Path_to_.NET_Framework\MSBuild.exe /property:Configuration=Release nanodb.csproj

Q: Is nanodb.exe portable?
A: No, this working with files from restore-archive.

## Updates

See updates here: https://github.com/username1565/nanoboard/releases/

Commits: here: https://github.com/username1565/nanoboard/network


Changes.txt
Changes:
File .../nanoboard-restore/nanodb.exe-source/PngTransport/Aggregator.cs:
	-	Add code to get the "host_and_folder" "protocol://(domain/IP):PORT/folder"
		and get this from the thread links, like this: "https://0chan.ml/_frct/res/1.html?query=some_query&parameter=value"
	
	-	Get (domain/IP):PORT only (variable "host"), and protocol://(domain/IP):PORT/ without folder (variable "protocol_host_port") - too.
		Now IP:port can be used for threads in "Settings". This will be parsed correctly, if path of PNG is relative, in the response from server.
	
	-	Do this once for each thread, not for each picture.
	
	-	Add regular expression to get picture_host from imageAddress(url in parsed "href") if this contains "http://" or "https://"
	
	-	Block downloading pictures using unknown protocol. For example, "ftp://domain.com/picture.png" in "href" - will be blocked.
	
	-	Change imageAddress by adding pictures relative path from href:
		to "protocol_host_port" (if in href "/images/pic.png" or "./images/pic.png")
		to "host_and_folder" (if in href "../images/pic.png" or "images/pic.png").
	
	-	Add IP-logging attempt prevention:
		if imageAddress contains keyword "logger", or URL shotrers "bit.ly" and "goo.gl" - picture downloading blocked.
		if picture host not thie host of the thread (hard prevention).
		There is possible to comparing protocols+host, because IP-loggers can working on another protocols,
		but if HTTPS board load images using HTTP, this prevention not good.
	
	-	Change m-2ch.ru to m2ch.hk.
	
	-	Add description and second condition as operand for prevent IP-logging, but load pictures from "8ch.net (board)" -> "media.8ch.net (images)".
	
	-	At line 298: Add re-calculating the picture_host for each imageAddress, because many png's from dobrochan
		was been blocked, after downloading PNG from iplogger.org - blocked...

File: .../nanoboard-restore/nanodb.exe-source/Server/DbApiHandler.cs
	-	At line 131: the function was been added parse_number(string string_number), because int.Parse and Int32.Parse not working correctly for me.
	-	At line 150-151: replaced using this. Now last 50 and las 500 messages working good.
	
_____________________________________________________________________________________________________________________________________________

Changes in nanoboard 3.2
_____________________________________________________________________________________________________________________________________________
Short changes:
	-	change style small fix.
	- 	add show-hide post by click on eye. Like on 2ch.hk.
		Two strings from makaba.css used in the tag <style> in index.html.
		Functions added in nanoclient.js and can be founded by keyword "post_type_hidden".
		Post hash saved in LocalStorage.

	-	add separate tab to create PNG.
	-	add queue there. Now you can pack specified or selected posts and customize randomizing packing.
	-	Add queue button for each post. Now by click to this button any post can be added and removed from queue.
	-	Add page to unpack custom container. 	Now this located by http://127.0.0.1:7346/pages/Open_container.html
	-	Add page to convert images to PNG.		Now this located here: http://127.0.0.1:7346/pages/convert-to-PNG.html
	-	Add PNG generator. Bitmap or FractalGen. DataURL sent by GET or POST-query (for big images).
		See method generate_PNG in /Server/DbApiHandler.cs
	-	Add more GET and POST-parameters for api-methods there. 
	-	Update post count /count - now only when collect.
	-	Update places not by interval, now. There is separate link to update this by click.
	-	notif - update by long interval, but when notif, this swithed to short interval. See scripts/params.js
	-	Add check sha256 hash of the file captcha.nbc. This is checking once after try to load captcha. So no need captcha.nbc for readonly mode.
		URL for downloading capcha is not hardcoded, and can be changed to custom URL - in config-3.json.
	-	Add static captcha hash to config-3.json and this value is hardcoded in the code. See nanoboard-source/Captcha/Captcha.cs
	-	Captcha not downloaded if fail, because:
		1. Github using regirect.
		2. This is not anonymous action.
		Need to add proxy, so code to downloading captcha was been commented.
	-	Proxy! This was been added to collectPNG.
		See AddProxy method in /PngTransport/DbApiHandler.cs
		/proxy.txt - list of proxies + comment.
		When collect, each proxy from this list try to ping, then first successfully proxy using as default, and pictures is downloading using this proxy.
	-	Change post validation to bypassValidation, when captcha.nbc not found. No error now.
	-	Add all changes from nanoboard 3.1. See thread http://127.0.0.1:7346/pages/index.html#thread70ca25ba04a241d3b3729611205bf4bb
		All changes from [3.1 version](https://github.com/username1565/nanoboard/commit/e828b228af6d2a1cc507ff80a59c25a0f5a511f0)
		According the comment https://github.com/username1565/nanoboard/issues/3#issuecomment-463904141 from issue https://github.com/username1565/nanoboard/issues/3
		3 level for replyes, when thread loaded from main page.
		Sorting last posts by date.
		Bump for threads.
		Add comparer to make avaiable to compile version 3.1 in .NET Framework 3.5. See post 082a06592e3b241e3c49063524a3ac25
_____________________________________________________________________________________________________________________________________________
	Full changes:
				\Changes.txt			this file - changed. This info added.
				\pages\index.html		Just change version to 3.1
				\pages\version.txt		Just change version to 3.1
				\scripts\nanoclient.js	add function AddBriefView from 3.1
				\scripts\reply.js		Just change version to 3.1
				
				\nanodb.exe-source\nanodb.csproj
				\nanodb.exe-source\nanodb.sln
				\nanodb.exe-source\Program.cs
						copy this 3 files from 3.1 client with changes.
				\nanodb.exe-source\app.config
				\nanodb.exe-source\nanodb.csproj.user

scripts/last.js

function loadRootThread(hash) {
  var first = hash;
  var prev = hash;
  var fn = function(){
    $.get('../api/get/' + hash)
      .done(function(p){
        p = JSON.parse(p);
        if (p.replyTo == _categories)
        {
          _depth = 2;
          loadThread(prev, first);
          return;
        }
        prev = p.hash;
        hash = p.replyTo;
        fn();
      });
  };
  fn();
}

function loadRootThreadHash(hash,pp,cb) {
  $.post('../api/find-thread/' + hash, _categories)
    .done(function(res){
      cb(res,pp,hash);
    })
    .fail(function(){
      cb(null,pp,hash);
    });
/*
  var first = hash;
  var prev = hash;
  var fn = function(){
    $.get('../api/get/' + hash)
      .done(function(p){
        p = JSON.parse(p);
        if (p.replyTo == _categories)
        {
          cb(prev,pp);
          return;
        }
        prev = p.hash;
        hash = p.replyTo;
        fn();
      });
  };
  fn();
*/
}

function showLast(N, from_index){
	
	//console.log('from_index', from_index, 'N', N);
	
	var from_index = (typeof from_findex === 'undefined') ? false : from_findex;
	
	//console.log('from_index', from_index, 'N', N);
    
//    $.get('../api/pcount')
    $.get('../api/count')
      .done(function(cnt){
        if(from_index===false){
			cnt = parseInt(cnt);
			total_posts.innerHTML = cnt;

			var si_div = start_index_last_posts_div;	//get hidden div element
			si_div.style.display = "block";				//show div with input element
			var lsi = lastli_start_index;				//get input element with for start_index

			from_index =
			lsi.value = 
						( (cnt-N) < lsi.value )
							? Math.max(cnt-N,0)			//set value of start index
							: lsi.value					//or previous value
			;

			lsi.min = 0;								//only positive numbers.
			lsi.max = cnt;								//up to post count
			showN.max = cnt;							//no more than post count...
			showN.value = (N>cnt) ? cnt : N;			//Set value, if selected 500 or 200 posts, when total post count lesser this values.
			
		}else{
			from_index = Math.max(from_index,0);								//if from_index is negative number this will be 0
			//showN.value = (N>(cnt-from_index)) ? (cnt-from_index) : N;		//Set value			
			showN.value = (N>(cnt-from_index)) ? (cnt-from_index) : from_index;	//Set value			
		}
		
/*
		console.log(
					"function showLast: "
		,	'\n',	"from_index", from_index
		//,	'\n',	"si_div.style.display", si_div.style.display
		//,	'\n',	"lsi", lsi
		//,	'\n',	"lsi.value", lsi.value
		);
*/

		if( ( from_index === '' ) || ( [ 10, 50, 100, 200, 500 ].indexOf( N )!==-1 ) ){
			from_index = lsi.value = ( ( ( cnt - N ) < 0 ) ? 0 : ( cnt - N ) );
		}

		//console.log('from_index', from_index, 'N', N, 'cnt', cnt);
    
//        $.get('../api/prange/'+from_index+'-'+N+'-'+'true')					//start_index = cnt-N || 0 - not negative.	//last parameter appendText = true, to append text info if post was been reported.
//        $.get('../api/prange/'+from_index+'-'+N+'-'+'false')					//start_index = cnt-N || 0 - not negative.	//last parameter appendText = true, to append text info if post was been reported.

        $.get('../api/prange/'+from_index+'-'+N+'-'+'append_text'+'-'+'show_deleted')	//start_index = cnt-N || 0 - not negative.	//last parameter appendText = true, to append text info if post was been reported.

          .done(function(arr){
            active_tab("lastli")
            arr = JSON.parse(arr);
/*
//3.1
			if(document.getElementById('createPNG').style.display === 'block'){
				var threadId = 'thread';
			}else{
				var threadId = 'queue_thread';
			}
*/			
            if (arr.length > 0) {
//3.0
              $('#thread').empty();
//3.1
//              $('#'+threadId).empty();
            } else {return;}
            for (var i = arr.length - 1; i >= 0; i--) {
//3.0
				if(arr[i]==null){continue;} //skip null, if post was been deleted.
				
              var p = addPost(arr[i], function(d) { d.appendTo($('#thread')); }, false);
				
//				if (arr[i].message === 'cG9zdCB3YXMgZGVsZXRlZA=='){
				if ((arr[i].message).substring(0, 16) === 'post_was_deleted'){
					p.css({ opacity: _deletedOpacity});
				}
//3.1
//              var p = addPost(arr[i], function(d) { d.appendTo($('#'+threadId)); }, false);
				if (
						arr[i].hash != _categories
					&&	arr[i].replyTo != _categories
					&&	arr[i].replyTo != _rootpost
					&&	p
				){
				  
					//if post is reply for existing root thread
					loadRootThreadHash(
						p.attr('id'),
						p,
						function(h,pp,post_hash) {
							pp.append(
								$('<a>')
//3.0
								.attr('href', '#thread' + ( (h == null) ? post_hash : h ) )
//3.1
//              		        .attr('href', '#'+threadId + h)
								.html(
									'<span class="glyphicon glyphicon-menu-hamburger" aria-hidden="true"></span>'+
									'<span class="btn-title">&thinsp;'
									+
										(h == null ? 'Thread Not Found' : 'Thread')
									+
									'</span>'
								)
								.click(function(){
									//loadRootThread($(this).parent().attr('id'));
								})
							);
					
							pp.append(
								$('<a>')
								.attr('href', '#thread' + post_hash)
								.html(
									'&nbsp;&nbsp;&nbsp;<span class="glyphicon glyphicon-paperclip" aria-hidden="true"></span>'+
									'<span class="btn-title">&thinsp;POST</span>'
								)
								.click(function(){
									//loadRootThread($(this).parent().attr('id'));
								})
							);
						}
					);
				}else{
					//if post is not reply for existing root thread, and root thread not found
					
					//Append "Thread not found"
					$('#'+arr[i].hash).append(
						$('<a>')
							.attr('href', '#thread' + arr[i].hash )
							.html(
								'<span class="glyphicon glyphicon-menu-hamburger" aria-hidden="true"></span>'+
								'<span class="btn-title">&thinsp;'
								+'Thread Not Found'+
								'</span>'
							)
							.click(function(){
								//loadRootThread($(this).parent().attr('id'));
							})
					);

					//Append "Link to single post"
					$('#'+arr[i].hash).append(
						$('<a>')
						.attr('href', '#thread' + arr[i].hash)
						.html(
							'&nbsp;&nbsp;&nbsp;<span class="glyphicon glyphicon-paperclip" aria-hidden="true"></span>'+
							'<span class="btn-title">&thinsp;POST</span>'
						)
						.click(function(){
							//loadRootThread($(this).parent().attr('id'));
						})
					);
				}
            }
            vid_show()
          });
      });
  }
